File: lang/go/src/fun/vac/algorithms/sorting/QuickSort/QuickSort.go
package QuickSort

func Partition(arr []int, lo int, hi int) {
    if lo == hi {
        return
    }

    pivot := arr[lo]

    left, right := lo, hi - 1

    for left != right {
        for i := right; i > left; i-- {
            if arr[i] < pivot {
                arr[left] = arr[i]
                arr[i] = pivot
                break
            }
            right -= 1
        }
        for i := left; i < right; i++ {
            if arr[i] > pivot {
                arr[right] = arr[i]
                arr[i] = pivot
                break
            }
            left += 1
        }
    }

    Partition(arr, lo, left)
    Partition(arr, left + 1, hi)
}

func Sort(arr []int) {
    Partition(arr, 0, len(arr))
}

File: archive/php/muse/algorithms/sorting/QuickSort.php
<?php

declare(strict_types=1);

namespace muse\algorithms\sorting\QuickSort;

function partition(array &$arr, int $lo, int $hi): void
{
    if ($lo === $hi) {
        return;
    }

    $pivot = $arr[$lo];

    $left  = $lo;
    $right = $hi - 1;

    while ($left != $right) {
        for ($i = $right; $i > $left; $i--) {
            if ($arr[$i] < $pivot) {
                $arr[$left] = $arr[$i];
                $arr[$i] = $pivot;
                break;
            }
            $right--;
        }
        for ($i = $left; $i < $right; $i++) {
            if ($arr[$i] > $pivot) {
                $arr[$right] = $arr[$i];
                $arr[$i] = $pivot;
                break;
            }
            $left++;
        }
    }

    partition($arr, $lo, $left);
    partition($arr, $left + 1, $hi);
}

function sort(array &$arr): void
{
    partition($arr, 0, count($arr));
}
